### PR TITLE
feat: :sparkles: Added filtering option to FAISS vectorstore

### PR DESCRIPTION
**Pull request by vempaliakhil96**
Inspired by the filtering capability available in ChromaDB, added the same functionality to the FAISS vectorestore as well. Since FAISS does not have an inbuilt method of filtering used the approach suggested in this [thread](https://github.com/facebookresearch/faiss/issues/1079)
Langchain Issue inspiration: https://github.com/hwchase17/langchain/issues/4572

- [x] Added filtering capability to semantic similarly and MMR
- [x] Added test cases for filtering in `tests/integration_tests/vectorstores/test_faiss.py`

#### Who can review?

Tag maintainers/contributors who might be interested:

  VectorStores / Retrievers / Memory
  - @dev2049
  - @hwchase17

### CODE REVIEW

File: langchain/vectorstores/faiss.py [hunk 1]
```diff
@@ -180,13 +180,19 @@ def add_embeddings(
...
+        filter: Optional[Dict[str, str]] = None,
+        fetch_k: int = 20,
     ) -> List[Tuple[Document, float]]:
         """Return docs most similar to query.
 
         Args:
             embedding: Embedding vector to look up documents similar to.
             k: Number of Documents to return. Defaults to 4.
+            filter (Optional[Dict[str, str]]): Filter by metadata. Defaults to None.
+            fetch_k: Number of Documents to fetch after filtering
```
Comment (hwchase17) [inline langchain/vectorstores/faiss.py:195]: this should read "before filtering" right?

**Review by hwchase17 — COMMENTED** [review]
this is fantastic! can we add an example to the FAISS notebook?

File: langchain/vectorstores/faiss.py (hunk 1 above)
Comment (vempaliakhil96) [reply langchain/vectorstores/faiss.py:195]: Ohh yes. I will edit and add examples in docs as well. Thanks!

**Review by hwchase17 — APPROVED** [review]
lgtm - thanks!

File: langchain/vectorstores/faiss.py (hunk 1 above)
Comment (TasksAndreySorokin) [reply langchain/vectorstores/faiss.py]: Now it is possible to filter after docs are found! Is it possible to filter first by metadata and then look for a match to the request?

Comment (JasonChaa) [reply langchain/vectorstores/faiss.py]: I'm looking for that too

**Pull request merged**